File: sdks/node/test/core/admission.test.ts
// S26 — opt-in `maxPending` admission cap. Jobs stay pending without a worker,
// so the cap is exercised purely producer-side.

import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { expect, it } from "vitest";
import { Queue, QueueError, QueueFullError } from "../../src/index";

function newQueue(): Queue {
  const dbPath = join(mkdtempSync(join(tmpdir(), "taskito-adm-")), "queue.db");
  return new Queue({ dbPath });
}

it("countPendingByQueue counts pending per queue", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  expect(queue.countPendingByQueue("default")).toBe(0);
  queue.enqueue("noop");
  queue.enqueue("noop");
  expect(queue.countPendingByQueue("default")).toBe(2);
});

it("uncapped queue never rejects", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  for (let i = 0; i < 25; i++) queue.enqueue("noop");
  expect(queue.countPendingByQueue("default")).toBe(25);
});

it("rejects at the configured cap", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  queue.configureQueue("default", { maxPending: 2 });
  queue.enqueue("noop");
  queue.enqueue("noop");
  expect(() => queue.enqueue("noop")).toThrow(QueueFullError);
  // Rejected enqueue inserted nothing.
  expect(queue.countPendingByQueue("default")).toBe(2);
});

it("cap is per queue", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  queue.configureQueue("tight", { maxPending: 1 });
  queue.enqueue("noop", undefined, { queue: "tight" });
  expect(() => queue.enqueue("noop", undefined, { queue: "tight" })).toThrow(QueueFullError);
  for (let i = 0; i < 5; i++) queue.enqueue("noop", undefined, { queue: "wide" });
});

it("enqueueMany is all-or-nothing against the cap", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  queue.configureQueue("default", { maxPending: 3 });
  queue.enqueue("noop");
  queue.enqueue("noop");
  queue.enqueue("noop");
  expect(() => queue.enqueueMany("noop", [{}, {}, {}])).toThrow(QueueFullError);
  expect(queue.countPendingByQueue("default")).toBe(3);
});

it("enqueueMany accounts for the batch size", () => {
  const queue = newQueue();
  queue.task("noop", () => undefined);
  queue.configureQueue("default", { maxPending: 3 });
  // Empty queue, but a batch bigger than the cap is rejected as a whole.
  expect(() => queue.enqueueMany("noop", [{}, {}, {}, {}])).toThrow(QueueFullError);
  expect(queue.countPendingByQueue("default")).toBe(0);
  // A batch that exactly fits is admitted.
  queue.enqueueMany("noop", [{}, {}, {}]);
  expect(queue.countPendingByQueue("default")).toBe(3);
  // Now full: one more is rejected.
  expect(() => queue.enqueue("noop")).toThrow(QueueFullError);
});

it("configureQueue rejects a negative cap", () => {
  const queue = newQueue();
  expect(() => queue.configureQueue("default", { maxPending: -1 })).toThrow(RangeError);
});

it("QueueFullError is a QueueError", () => {
  const err = new QueueFullError("q", 5, 5);
  expect(err).toBeInstanceOf(QueueError);
  expect(err.queue).toBe("q");
  expect(err.cap).toBe(5);
});

File: sdks/java/src/test/java/org/byteveda/taskito/core/AdmissionTest.java
package org.byteveda.taskito.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.nio.file.Path;
import org.byteveda.taskito.Taskito;
import org.byteveda.taskito.errors.QueueFullException;
import org.byteveda.taskito.task.EnqueueOptions;
import org.byteveda.taskito.task.Task;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.io.TempDir;

/** S26 — opt-in {@code maxPending} admission cap. No worker runs, so jobs stay pending. */
class AdmissionTest {

    private static Taskito sqlite(Path dir) {
        return Taskito.builder()
                .backend("sqlite")
                .url(dir.resolve("t.db").toString())
                .open();
    }

    @Test
    @Timeout(30)
    void countsPendingPerQueue(@TempDir Path dir) {
        Task<String> noop = Task.of("noop", String.class);
        try (Taskito queue = sqlite(dir)) {
            assertEquals(0, queue.countPendingByQueue("default"));
            queue.enqueue(noop, "a");
            queue.enqueue(noop, "b");
            assertEquals(2, queue.countPendingByQueue("default"));
        }
    }

    @Test
    @Timeout(30)
    void uncappedQueueNeverRejects(@TempDir Path dir) {
        Task<String> noop = Task.of("noop", String.class);
        try (Taskito queue = sqlite(dir)) {
            for (int i = 0; i < 25; i++) {
                queue.enqueue(noop, String.valueOf(i));
            }
            assertEquals(25, queue.countPendingByQueue("default"));
        }
    }

    @Test
    @Timeout(30)
    void rejectsAtCap(@TempDir Path dir) {
        Task<String> noop = Task.of("noop", String.class);
        try (Taskito queue = sqlite(dir)) {
            queue.maxPending("default", 2);
            queue.enqueue(noop, "a");
            queue.enqueue(noop, "b");
            assertThrows(QueueFullException.class, () -> queue.enqueue(noop, "c"));
            // Rejected enqueue inserted nothing.
            assertEquals(2, queue.countPendingByQueue("default"));
        }
    }

    @Test
    @Timeout(30)
    void enqueueManyAccountsForBatchSize(@TempDir Path dir) {
        Task<String> noop = Task.of("noop", String.class);
        try (Taskito queue = sqlite(dir)) {
            queue.maxPending("default", 3);
            // Empty queue, but a batch bigger than the cap is rejected as a whole.
            assertThrows(
                    QueueFullException.class, () -> queue.enqueueMany(noop, java.util.List.of("a", "b", "c", "d")));
            assertEquals(0, queue.countPendingByQueue("default"));
            // A batch that exactly fits is admitted.
            queue.enqueueMany(noop, java.util.List.of("a", "b", "c"));
            assertEquals(3, queue.countPendingByQueue("default"));
            // Now full: one more is rejected.
            assertThrows(QueueFullException.class, () -> queue.enqueue(noop, "x"));
        }
    }

    @Test
    void rejectsNegativeCap(@TempDir Path dir) {
        try (Taskito queue = sqlite(dir)) {
            assertThrows(IllegalArgumentException.class, () -> queue.maxPending("default", -1));
        }
    }

    @Test
    @Timeout(30)
    void capIsPerQueue(@TempDir Path dir) {
        Task<String> noop = Task.of("noop", String.class);
        try (Taskito queue = sqlite(dir)) {
            queue.maxPending("tight", 1);
            EnqueueOptions tight = EnqueueOptions.builder().queue("tight").build();
            queue.enqueue(noop, "a", tight);
            assertThrows(QueueFullException.class, () -> queue.enqueue(noop, "b", tight));
            // A different, uncapped queue is unaffected.
            EnqueueOptions wide = EnqueueOptions.builder().queue("wide").build();
            for (int i = 0; i < 5; i++) {
                queue.enqueue(noop, String.valueOf(i), wide);
            }
        }
    }
}
